### PR TITLE
Fix parsing of subscript expressions using python resolver

### LINKED ISSUE #38728: [JIT] Finish Resolving Python Types By Value
## 🐛 Bug

To recursively compile python programs, we need to recursively look up the python types that are used. We do this in many locations currently, but not all. 

Something like:
```
class X(object):
    def __init__(self):
        pass

@torch.jit.script
def foo():
    return X()
    pass
```
Will correctly resolve `X()` and compile , but if you modify the function to be:
```
@torch.jit.script
def foo(out: X):
    return out
```
 
it will fail with `Unknown type name 'X':`. This is because in many locations in the compiler we try to resolve the code (e.g. [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/ir_emitter.cpp#L1545)).  The textual resolution is needed for compilation in C++, but when the python environment is available we should search it using `resolveType` [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/ir_emitter.cpp#L503). We will need to handle cases that use `.` and `[]`, such as `List[int]`, and `namespace.X`. 

This has been an issue for a number of users, and is a particular issue with `torch.jit.Future`. `torch.jit.Future` is the runtime type value,  but if users use `torch.jit.Future` as an inline type annotation it will fail.

The first part of this issue was https://github.com/pytorch/pytorch/pull/29623/files. This issue may require multiple PRs to completely fix, and you may run into other issues when doing it.

### PR DESCRIPTION
**Pull request by wconstab**
- add call out to python resolver in parseArgsFromDecl, parserReturnFromDecl
- add support in python resolver for nested subexpressions
- wrap python resolver call in exception handling to fall back to c++ path
- add tests for newly resolvable types 
- closes #38728

Fixes bug where SourceRange objects did not include the final closing ']' for a subscript expression.  E.g. range for 'List[int]' previously included only 'List[int'.

### CODE REVIEW

File: torch/_jit_internal.py [hunk 1]
```diff
@@ -31,12 +31,50 @@ def env(qualified_name, module):
...
+            # assume only subexp (between []) could contain ','
+            # but allow each element of the list to be any type hence recursive env()
+            subexp = qualified_name[first + 1: last]
+            if ',' in subexp and (subexp.find('[') == -1 or
+                                  subexp.find('[') > subexp.find(',')):
+                no_whitespace = "".join(subexp.split())
+                parts = no_whitespace.split(',')
+                types = [env(p, module) for p in parts]
+
+                # HACK for now - must be a better way to approximate base_mod[*types]
```
Comment (wconstab) [inline torch/_jit_internal.py:52]: @eellison Any suggestion for a better way to do this part?

File: torch/_jit_internal.py [hunk 2]
```diff
@@ -31,12 +31,50 @@ def env(qualified_name, module):
...
+                # either base or subexp could include '.' or '['
+                subexp_mod = env(subexp, module)
+                return base_mod[subexp_mod]
+        elif '.' in qualified_name:
             parts = qualified_name.split('.')
             base = parts[0]
             remainding_pieces = '.'.join(parts[1:])
             module_value = getattr(module, base)
             return env(remainding_pieces, module_value)
+        elif 'int' == qualified_name:
```
Comment (wconstab) [inline torch/_jit_internal.py:74]: thoughts on doing this here? or should these type of things result in a bailout back to cpp too?

if this is the right approach, then i'd probably want to be more thorough about the types we handle.

File: torch/csrc/jit/frontend/script_type_parser.cpp [hunk 3]
```diff
@@ -162,13 +162,22 @@ c10::optional<std::string> ScriptTypeParser::parseBaseTypeName(
 
 TypePtr ScriptTypeParser::parseTypeFromExpr(const Expr& expr) const {
   if (expr.kind() == TK_SUBSCRIPT) {
+
+    if (resolver_) {
```
Comment (wconstab) [inline torch/csrc/jit/frontend/script_type_parser.cpp:166]: I put this in as a hack at first with a hardcoded string passed in.  using expr.range().text() seems to be correct in most cases. I'm not sure how to decide if it's actually an OK approach or still a hack.

**Review by eellison — COMMENTED** [review]
this is looking really good ! A few small fixes we need to make but I think this is close. 

also mind updating the title of the PR and adding some context to it (maybe link to original issue)

File: test/test_jit_py3.py [hunk 4]
```diff
@@ -403,6 +403,40 @@ def wrong_type():
...
+
+        g = fn.graph
+        assert "Future" in repr(g)
+
+    def test_subexpression_Optional(self):
+        assert False, "TODO implement me"
+
+    def test_unimported_type_resolution(self):
+        # verify fallback from the python resolver to the c++ resolver
+        assert False, "TODO implement me, maybe in a new file?"
```
[the remark this answers is not in the document]
Comment (wconstab) [reply test/test_jit_py3.py:445]: oh, yea that's a good workaround; and I guess I can first assert in the test that the type isn't resolved, to guard against someone else importing it in the future.  I was overthinking it perhaps but wanted to handle things that already were imported, which i don't think i can do in the same file but maybe can do in another file...

File: torch/_jit_internal.py (hunk 2 above)
Comment (wconstab) [reply torch/_jit_internal.py:64]: i'll come back to this one; i only added it to fix some failures in running test_jit.py, and in those cases getattr(module, 'int') was failing for me i thought.

File: test/test_jit_py3.py [hunk 5]
```diff
@@ -403,6 +403,40 @@ def wrong_type():
...
             torch.jit.script(wrong_type)
 
+    def test_subexpression_List_Future(self):
+
+        @torch.jit.script
+        def fn(x: List[torch.jit.Future[int]]):
+            return x
+
+        g = fn.graph
+        assert "Future" in repr(g)
```
[the remark this answers is not in the document]
Comment (wconstab) [reply test/test_jit_py3.py]: cool. maybe for the optional case i can use check-dag? not sure how to best use it in that case.

File: test/test_jit_py3.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wconstab) [reply test/test_jit_py3.py]: ok i sorta made this test but i left it WIP.  Is the point just to test that it runs?  or is it supposed to invoke my code? i was also surprised that 'number' got implicitly converted to tensor, so i was hesitant to assert for Tensor in the FileCheck part.

File: torch/_jit_internal.py (hunk 2 above)
Comment (wconstab) [reply torch/_jit_internal.py]: yea these were able to go away.  i had added them in to good effect, they made a test pass. but taking them out and adding in more thorough 'fallback' (return None) coverage fixed it too.

File: torch/_jit_internal.py [hunk 6]
```diff
@@ -26,17 +26,81 @@ def createResolutionCallbackFromEnv(lookup_base):
...
+            return base_mod[subexp_mod]
+        except TypeError:
+            return None
+
+    def subexpression(subexp, base_mod, module):
+        if ',' in subexp:
+            subexp = "".join(subexp.split())
+            splits = []
+            nest_level = 0
+            for i, c in enumerate(subexp):
```
Comment (wconstab) [inline torch/_jit_internal.py:40]: thoughts on this, cleaner/easier way to do it?

File: torch/_jit_internal.py [hunk 7]
```diff
@@ -26,17 +26,81 @@ def createResolutionCallbackFromEnv(lookup_base):
...
         # over scope, then proceed to use the looked-up value to go down the
         # chain.
-        if '.' in qualified_name:
+        if '[' in qualified_name:
+            first, last = qualified_name.find('['), qualified_name.rfind(']')
+            base = qualified_name[:first]
+            base_mod = env(base, module)
+
+            # intentional bailout, e.g. base == "Tuple" but wasn't imported in Module
+            if not base_mod or isinstance(base_mod, torch.Tensor):
```
Comment (wconstab) [inline torch/_jit_internal.py:90]: oops, the part after 'or' was meant to be cleaned up. removing it after retesting..

File: torch/_jit_internal.py [hunk 8]
```diff
@@ -26,17 +26,81 @@ def createResolutionCallbackFromEnv(lookup_base):
     You should not use this directly, it should only be used from the other
     createResolutionCallbackFrom* functions.
     """
+    def try_build_type(base_mod, subexp_mod):
```
Comment (wconstab) [inline torch/_jit_internal.py:29]: i started having logic in multiple places and it was still failing in cases like a unit test that tries Tensor[Tensor] and expects a particular exception.

**Review by eellison — COMMENTED** [review]
The current parsing logic is kind of complicated / hard to follow. I'm not sure exactly if if it works robustly (although the test cases look good). I have one suggestion of how to make it easier but looks almost ready to land.

File: torch/_jit_internal.py [hunk 9]
```diff
@@ -26,17 +26,76 @@ def createResolutionCallbackFromEnv(lookup_base):
...
+                if None in types:
+                    return None
+                return try_build_type(base_mod, tuple(types))
+
+            else:
+                subexp_mod = env(subexp, module)
+                return try_build_type(base_mod, subexp_mod)
+        else:
+            # either base or subexp could include '.' or '['
+            if 'Tensor' == subexp:
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/_jit_internal.py:73]: yea, i think that's what this was from.  i'd only point out that i think it was failing in a way that was not gracefully falling back to the old parser, so i had to do something about it (either make it work, or make it return None).  I've leaned more toward the latter in general, but i'll have to revisit this case and see if i can just delete it now that I have checks/try/except in more places or if i'll still need some specific handling for it.

File: test/test_jit_py3.py [hunk 10]
```diff
@@ -403,6 +403,51 @@ def wrong_type():
...
+
+    def test_unimported_type_resolution(self):
+        # verify fallback from the python resolver to the c++ resolver
+
+        @ torch.jit.script
+        def fn(x):
+            # type (number) -> number
+            return x + 1
+
+        FileCheck().check('number').run(fn.graph)
```
Comment (wconstab) [inline test/test_jit_py3.py:449]: @eellison Clarification on what you were wanting this test to cover and what the expected behavior of the system is (i observed that the number gets converted to a Tensor which initially surprised me but may be expected - should I assert on Tensor?)

Comment (eellison) [reply test/test_jit_py3.py:449]: numbers gets converted to a tensor with `x + 1` ? that is weird.

**Review by eellison — APPROVED** [review]
This is awesome! Definitely a tricky PR, but is much needed to release `torch.jit.fork`. This will also make it super easy to handle https://docs.python.org/3.9/whatsnew/3.9.html#pep-585-builtin-generic-types without adding weird special handling to our script type parser. 

I think this also fixes: https://github.com/pytorch/pytorch/issues/16492 and https://github.com/pytorch/pytorch/issues/32544 and

File: torch/_jit_internal.py [hunk 11]
```diff
@@ -26,21 +26,48 @@ def createResolutionCallbackFromEnv(lookup_base):
...
+            return base, i
+
+        assert expr[i] == '['
+        parts = []
+        while expr[i] != ']':
+            part_len = 0
+            i += 1
+            part, part_len = parseNestedExpr(expr[i:], module)
+            parts.append(part)
+            i += part_len
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/_jit_internal.py:56]: yea. ugh. this re-breaks my fix in the c++ side parser i think. taking a look..
